INJIMOB-3599] add authorize auth request in map and support for iar_post and iar_post.jwt

abhip2565 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Support for iar_post and iar_post.jwt response modes; accept authorization requests as Map inputs; new public APIs to construct VP responses and authorization error info.

* **Behavior**
  * IAR modes treated as valid aliases for existing handlers; authorization request state preserved and propagated; unsigned-request acceptance allowed when client validation is skipped.

* **Tests**
  * Expanded unit tests for response modes, Map requests, response/error construction, encryption and signing flows.

* **Documentation**
  * README clarified client-id scheme and unsigned-request rules.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KiruthikaJeyashankar committed:
[INJIMOB-3599] add: authorize auth request in map

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] add: constructVPResponse method to create success auth response

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] add: constructAuthorizationErrorResponse to create error auth response

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] fix: failing tests

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: change method anmes for constructing VP response and error

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: introduce response modes - iar_post, iar_post.jwt

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: remove unused responseMode alias param

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] test: add unit tests - constructVPToken, constructErrorInfo

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: extract duplicate to fucntion

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: modify error response to be map in direct_post.jwt

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: optimize imports

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/authorizationResponse/AuthorizationErrorResponse.kt:
```diff
@@ -0,0 +1,22 @@
+
+import com.google.gson.annotations.SerializedName
+import io.mosip.openID4VP.authorizationResponse.presentationSubmission.PresentationSubmission
+import io.mosip.openID4VP.authorizationResponse.vpToken.VPTokenType
```
Updated

KiruthikaJeyashankar commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/authorizationResponse/AuthorizationErrorResponse.kt:
```diff
@@ -0,0 +1,22 @@
+import io.mosip.openID4VP.authorizationResponse.vpToken.VPTokenType
+import io.mosip.openID4VP.common.encodeToJsonString
+
+private val className: String = AuthorizationErrorResponse::class.simpleName!!
```
Updated

KiruthikaJeyashankar commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/authorizationResponse/AuthorizationResponseHandler.kt:
```diff
@@ -123,6 +142,7 @@ internal class AuthorizationResponseHandler {
             )
         }
         try {
+            //TODO: make use of constructAuthorizationErrorResponse method - error would be in JWE/encrypted format ot not based on response_mode
```
ignore for now

KiruthikaJeyashankar replied:
NOt required , we are tracking this card

KiruthikaJeyashankar replied:
resolve this conversation

KiruthikaJeyashankar commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/responseModeHandler/types/DirectPostJwtResponseModeHandler.kt:
```diff
@@ -100,14 +120,31 @@ class DirectPostJwtResponseModeHandler : ResponseModeBasedHandler() {
+        val encryptedBody = jweHandler.generateEncryptedResponse(responseParams)
+
+        return mapOf("response" to encryptedBody)
     }
```
Updated

KiruthikaJeyashankar replied:
error responses need not be encrypted as of now

KiruthikaJeyashankar committed:
[INJIMOB-3599] test: add tests for finalizeAuthorizationResponse methods in response mode handler

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3576] add: allow unsigned request in pre-resgistered if client validation is false

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3576] test: fix failing tests

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3576] refactor:optimize imports

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3576] test: avoid force unwrapping for assertion

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3576] test: remove unused mocks

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar commented on kotlin/openID4VP/src/commonTest/kotlin/io/mosip/openID4VP/responseModeHandler/types/DirectPostJwtResponseModeHandlerTest.kt:
```diff
@@ -164,4 +176,159 @@ class DirectPostJwtResponseModeHandlerTest {
+    @Test
+    fun `finalizeAuthorizationResponse should encrypt AuthorizationErrorResponse successfully`() {
+        val walletNonce = "error-wallet-nonce"
+        val expectedEncryptedResponse = "encrypted.error.response"
+        val errorResponse = AuthorizationErrorResponse(
+            error = "invalid_request",
+            errorDescription = "Test error description",
+            state = "test-state"
+        )
+
+        every { anyConstructed<JWEHandler>().generateEncryptedResponse(any()) } returns expectedEncryptedResponse
+
+        val result = DirectPostJwtResponseModeHandler().finalizeAuthorizationResponse(
+            authorizationRequestForResponseModeJWT,
+            errorResponse,
+            walletNonce
+        )
+
+        assertEquals(
+            mapOf(
+                "error" to "invalid_request",
+                "error_description" to "Test error description",
+                "state" to "test-state"
+            ), result
+        )
+    }
```
updated

swatigoel commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/authorizationRequest/authorizationRequestHandler/types/PreRegisteredSchemeAuthorizationRequestHandler.kt:
```diff
@@ -74,7 +74,7 @@ class PreRegisteredSchemeAuthorizationRequestHandler(
         }
 
-        return false
+        return true
```
Need to update method documentation
`If client validation is disabled, then the request by value is not supported` This is not correct now.

KiruthikaJeyashankar replied:
Updated

swatigoel commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/constants/ResponseMode.kt:
```diff
@@ -3,4 +3,7 @@ package io.mosip.openID4VP.constants
 enum class ResponseMode(val value: String) {
     DIRECT_POST("direct_post"),
     DIRECT_POST_JWT("direct_post.jwt"),
+
```
any particular reason for new line?

KiruthikaJeyashankar replied:
Yeah , wanted to provide some separation b/w OVP and PDI flow's response modes

swatigoel commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/responseModeHandler/ResponseModeBasedHandler.kt:
```diff
@@ -41,4 +42,16 @@ abstract class ResponseModeBasedHandler {
+        authorizationResponse: AuthorizationErrorResponse,
+        walletNonce: String
+    ): Map<String, String>
 }
```
instead of having 2 methods, can we have one method where response can be either success or error?

swatigoel replied:
If we want to keep 2 separate methods then keep name also specific.
finalizeAuthorizationResponse, finalizeAuthorizationErrorResponse

KiruthikaJeyashankar replied:
merging to 1 method would cause issues as AuthorizationResponse & AuthorizationErrorResponse are different data clsses
Will refactor to have different names

KiruthikaJeyashankar replied:
updated

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: update docs

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: change method finalizeAuthorizationResponse to finalizeAuthorizationErrorResponse

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

swatigoel commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/responseModeHandler/types/DirectPostJwtResponseModeHandler.kt:
```diff
@@ -87,7 +88,44 @@ class DirectPostJwtResponseModeHandler : ResponseModeBasedHandler() {
+        )
+    }
+
+    override fun finalizeAuthorizationResponse(
```
finalizeAuthorizationResponse rename to getAuthorizationResponse

KiruthikaJeyashankar replied:
Updated

swatigoel commented on kotlin/openID4VP/src/commonMain/kotlin/io/mosip/openID4VP/responseModeHandler/types/DirectPostJwtResponseModeHandler.kt:
```diff
@@ -87,7 +88,44 @@ class DirectPostJwtResponseModeHandler : ResponseModeBasedHandler() {
+        )
+    }
+
+    override fun finalizeAuthorizationErrorResponse(
```
finalizeAuthorizationErrorResponse rename to getAuthorizationErrorResponse

KiruthikaJeyashankar replied:
Updated

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: rename ResponseModeBasedHandler method

 Changes
 - finalizeAuthorizationResponse & finalizeAuthorizationErrorResponse to getAuthorizationResponse & getAuthorizationErrorResponse

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] docs: update for unsigned request allowance

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3599] refactor: use getAuthorizationResponse in sendAuthorizationResponse

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>

KiruthikaJeyashankar committed:
[INJIMOB-3598] chore: update lib version to 0.7.0-SNAPSHOT

Signed-off-by: KiruthikaJeyashankar <kiruthikavjshankar@gmail.com>
